compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -360,7 +357,25 @@ where
     Qcx: QueryContext,
 {
     let state = Q::query_state(qcx);
-    match JobOwner::<'_, Q::Key, Qcx::DepKind>::try_start(&qcx, state, span, key) {
+    #[cfg(parallel_compiler)]
+    let state_lock = state.active.get_shard_by_value(&key).lock();
+    #[cfg(not(parallel_compiler))]
+    let state_lock = state.active.lock();
+
+    // For the parallel compiler we need to check both the query cache and query state structures
+    // while holding the state lock to ensure that 1) the query has not yet completed and 2) the
+    // query is not still executing. Without checking the query cache here, we can end up
+    // re-executing the query since `try_start` only checks that the query is not currently
+    // executing, but another thread may have already completed the query and stores it result
+    // in the query cache.
+    if cfg!(parallel_compiler) && qcx.is_sync() {
```

Comment:
The session is already accessible, no need for an extra method.

```suggestion
    if cfg!(parallel_compiler) && qcx.dep_context().sess().threads() > 1 {
```

r=me with the `is_sync` method removed.